packages/frontpage/lib/auth.ts

```diff
@@ -1,157 +1,547 @@
+      ? headers().get("host")
+      : process.env.VERCEL_ENV === "production"
+        ? process.env.VERCEL_PROJECT_PRODUCTION_URL!
+        : process.env.VERCEL_BRANCH_URL!;
+
+  const appUrl = `https://${host}`;
+
+  return {
+    // Client ID is the URL of the client metadata
+    // This isn't immediately obvious and if you supply something else the PAR request will fail with a 400 "Invalid url" error. I had to traverse the atproto implementation to find out why!
+    client_id: `${appUrl}/oauth/client-metadata.json`,
+    dpop_bound_access_tokens: true,
+    application_type: "web",
+    subject_type: "public",
+    grant_types: ["authorization_code", "refresh_token"],
+    response_types: ["code"], // TODO: "code id_token"?
+    scope: "openid profile offline_access",
+    client_name: "Frontpage",
+    token_endpoint_auth_method: "none",
+    redirect_uris: [`${appUrl}/oauth/callback`] as const,
+    client_uri: appUrl,
+    jwks_uri: `${appUrl}/oauth/jwks.json`,
+  } satisfies OAuthClientMetadata;
+});
+
+export const getOauthClientOptions = () =>
+  ({
+    client_id: getClientMetadata().client_id,
+    token_endpoint_auth_method: "private_key_jwt",
+  }) satisfies OauthClient;
+
+async function getClientPrivateKey() {
+  const jwk = await getPublicJwk().then(exportJWK);
+  const kid = jwk.kid;
+  if (!kid) {
+    // TODO: Fix this somehow? Maybe use webcrypto everywhere?
+    console.warn("Missing kid, fix this!");
   }
+  const key = await crypto.subtle.importKey(
+    "jwk",
+    await exportJWK(await getPrivateJwk()),
+    { name: "ECDSA", namedCurve: "P-256" },
+    true,
+    ["sign"],
+  );
+
+  return { key, kid };
 }
 
-const Credentials = z.object({
-  identifier: z.string(),
-  password: z.string(),
-});
+export async function signIn(handle: string) {
+  const did = await getDidFromHandleOrDid(handle);
+  if (!did) {
+    return {
+      error: "DID_NOT_FOUND",
+    };
+  }
 
-const auth = NextAuth({
-  providers: [
-    CredentialsProvider({
-      credentials: {
-        identifier: {},
-        password: {},
-      },
-      authorize: async (unsafeCredentials) => {
-        console.log("authorizing...");
-        const credentials = Credentials.parse(unsafeCredentials);
-        const session = await atprotoCreateSession({
-          password: credentials.password,
-          identifier:
-            // Remove @ from start if it's there
-            credentials.identifier.replace(/^@/, ""),
-        });
+  const meta = await getOauthResourceMetadata(did);
+  if ("error" in meta) {
+    return meta;
+  }
 
-        return {
-          id: session.did,
-          name: session.handle,
-          email: session.email,
-          refreshJwt: session.refreshJwt,
-          accessJwt: session.accessJwt,
-        };
+  const authServerUrl = meta.data.authorization_servers?.[0];
+  if (!authServerUrl) {
+    return {
+      error: "NO_AUTH_SERVER",
+    };
+  }
+
+  const authServer = await processDiscoveryResponse(
```

Comment:
Add schema checks on data returned from oauth4webapi where appropriate - it only checks for required properties in most cases but we should be more defensive